src/app/repositories/firebase/dataset_firebase_repository.py

```diff
@@ -31,44 +30,40 @@ def get_latest_dataset_with_survey_id(
             .stream()
         )
 
-    def create_new_dataset(
+    def perform_new_dataset_transaction(
         self,
         dataset_id: str,
         dataset_metadata_without_id: DatasetMetadataWithoutId,
-    ) -> None:
+        unit_data_collection_with_metadata: list[UnitDataset],
+        extracted_unit_data_rurefs: list[str],
+    ):
         """
-        Creates a new dataset in firestore with a specified ID and data.
+        Writes dataset metadata and unit data to firestore as a transaction, which is
+        rolled back if any of the operations fail.
 
         Parameters:
-        dataset_id (str): uniquely generated GUID id of the dataset.
-        dataset_metadata_without_id (DatasetMetadataWithoutId): metadata of the new dataset without id.
+        dataset_id: id of the dataset
+        dataset_metadata_without_id: dataset metadata without a dataset id
+        unit_data_collection_with_metadata: collection of unit data associated to a dataset
+        extracted_unit_data_rurefs: rurefs associated with the unit data collection
         """
-        logger.debug(
-            f"Setting dataset with id {dataset_id} and data {dataset_metadata_without_id}"
-        )
-        self.datasets_collection.document(dataset_id).set(dataset_metadata_without_id)
 
-    def get_dataset_unit_collection(self, dataset_id: str) -> list[UnitDataset]:
-        """
-        Gets the collection of units associated with a particular dataset.
+        @firestore.transactional
```

Comment:
Yeah I don't like it either, the problem is (as usual) the firestore sdk. This works as a single transaction because of the `@firestore.transactional` decorator. A stipulation of using this decorator is that the first parameter of the function has to be `transaction` or it throws an exception. However because our functions are class based, the first parameter is always `self`, so the only ways around it are either to move this function into a new file and call it from there, or have a function within a function.

I thought it was better to encapsulate it within the function because it isn't being used anywhere else, and if we moved this function to a separate file it wouldn't be consistent with the object orientated architecture of the rest of the application. Hope this makes sense, happy to discuss it further if not.

Reply:
Supplementing to Jack's answer:
https://stackoverflow.com/questions/58827943/getting-an-str-object-has-no-attribute-max-attempts-error-for-cloud-firesto
If we move the nested function out towards under the class object, the following error will be produced:
`2023-06-08 17:43:35,061 - services.schema.schema_processor_service - ERROR - Performing schema transaction: exception raised: 'SchemaFirebaseRepository' object has no attribute '_max_attempts'`

Reply:
Totally onboard on the rationale and the why of it. Could you just add a comment to the decorator function to explain thee rationale and why please. (Just so if we come back to it in the future, no on is second guessing why we did this pattern.